From c89657706803e3665c4fb8d9ba383ff68deb1520 Mon Sep 17 00:00:00 2001
From: Suman Anna <s-anna@ti.com>
Date: Thu, 15 Jan 2015 17:16:12 -0600
Subject: [PATCH 01/13] remoteproc/pruss: fix a crash with
 platform_device_unregister

The following kernel crash is seen during the removal of a PRUSS
device or during the removal of the PRUSS remoteproc module or
during any early error exit during the probe of the PRUSS device:

Unable to handle kernel NULL pointer dereference at virtual address 00000018
pgd = ec708000
[00000018] *pgd=ac6f8831, *pte=00000000, *ppte=00000000
Internal error: Oops: 17 [#1] SMP ARM
Modules linked in: pruss_remoteproc [last unloaded: virtio_rpmsg_bus]
CPU: 0 PID: 1164 Comm: sh Not tainted 3.14.26-00002-g49b864ea1130 #2
task: ec320040 ti: ec6dc000 task.ti: ec6dc000
PC is at release_resource+0x1c/0x90
LR is at release_resource+0x18/0x90
...[snip]...
Backtrace:
[<c004163c>] (release_resource) from [<c0394c6c>] (platform_device_del+0x74/0xa4)
 r4:00000001 r3:00000000
[<c0394bf8>] (platform_device_del) from [<c0394fa0>] (platform_device_unregister+0x14/0x20)
 r5:00000000 r4:ec1a4a00
[<c0394f8c>] (platform_device_unregister) from [<bf0071d8>] (pru_rproc_unregister+0x14/0x1c [pruss_remoteproc])
 r4:00000000 r3:ec1a4a10
[<bf0071c4>] (pru_rproc_unregister [pruss_remoteproc]) from [<c038f9a4>] (device_for_each_child+0x3c/0x7c)
[<c038f968>] (device_for_each_child) from [<bf007180>] (pruss_remove+0x40/0x84 [pruss_remoteproc])
 r6:ec1a4410 r5:ec247400 r4:ec247410
[<bf007140>] (pruss_remove [pruss_remoteproc]) from [<c0394914>] (platform_drv_remove+0x1c/0x20)
 r7:0000000f r6:c08b8ac0 r5:bf00954c r4:ec247410
[<c03948f8>] (platform_drv_remove) from [<c039308c>] (__device_release_driver+0x78/0xd0)
[<c0393014>] (__device_release_driver) from [<c0393108>] (device_release_driver+0x24/0x30)
 r5:ec247410 r4:ec247444
[<c03930e4>] (device_release_driver) from [<c03921d0>] (unbind_store+0x60/0x98)
 r5:bf00954c r4:ec247410
[<c0392170>] (unbind_store) from [<c03918fc>] (drv_attr_store+0x28/0x34)
 r7:0000000f r6:ec6ddf78 r5:ec6e8fc0 r4:ec70f900
[<c03918d4>] (drv_attr_store) from [<c0150c68>] (sysfs_kf_write+0x4c/0x50)
[<c0150c1c>] (sysfs_kf_write) from [<c015464c>] (kernfs_fop_write+0xbc/0x154)
 r5:00000000 r4:00000000
[<c0154590>] (kernfs_fop_write) from [<c00ee36c>] (vfs_write+0xb0/0x188)
 r10:b6fbb000 r9:ec6dc000 r8:0000000f r7:ec6ddf78 r6:b6fbb000 r5:0000000f
 r4:ec701980
[<c00ee2bc>] (vfs_write) from [<c00ee744>] (SyS_write+0x44/0x90)
 r10:b6fbb000 r8:0000000f r7:ec701980 r6:ec701980 r5:00000000 r4:00000000
[<c00ee700>] (SyS_write) from [<c000e940>] (ret_fast_syscall+0x0/0x30)
 r10:00000000 r8:c000eac4 r7:00000004 r6:0000000f r5:b6f945c8 r4:0000000f
Code: e1a04000 e59f0074 eb16dea3 e5943010 (e5932018)
---[ end trace 77712d823a300402 ]---

The crash is due to the differences between how the resources
are managed during the platform device creation path between
of_platform_populate() and the traditional platform_device_add().
The of_platform_populate() uses of_device_add() instead of
platform_device_add(), and as such does not add the resources.
The PRUSS device removal logic currently uses the traditional
platform_device_unregister() used to remove platform devices, and
this crashes during the cleanup of the resources. Fix this error
by using of_device_unregister(), the counterpart to of_device_add().

Signed-off-by: Suman Anna <s-anna@ti.com>
---
 drivers/remoteproc/pruss_remoteproc.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/remoteproc/pruss_remoteproc.c b/drivers/remoteproc/pruss_remoteproc.c
index abfe217..c16da32 100644
--- a/drivers/remoteproc/pruss_remoteproc.c
+++ b/drivers/remoteproc/pruss_remoteproc.c
@@ -1096,7 +1096,7 @@ static int pru_rproc_unregister(struct device *dev, void *data)
 {
 	struct platform_device *pdev = to_platform_device(dev);
 
-	platform_device_unregister(pdev);
+	of_device_unregister(pdev);
 
 	return 0;
 }
-- 
1.7.9.5

